firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+
+        Implemented in subclass."""
+        pass
+
+    @property
+    def operands(self):
+        """Returns the operands of the binary operation."""
+        return (self.children)
+
+    def __str__(self, prec=None):
+        """String representation of the binary operation."""
+
+        ops = {operator.add: '+',
+               operator.sub: '-',
+               operator.mul: '*'}
+        if prec is None or self.prec >= prec:
+            pars = lambda x: x
+        else:
+            pars = lambda x: "(%s)" % x
+        operand1 = self.children[0].__str__(prec=self.prec)
+        operand2 = self.children[1].__str__(prec=self.prec)
+        result = "%s %s %s" % (operand1, ops[self.op], operand2)
+
+        return pars(result)
+
+    def __repr__(self):
+        return "%s(%r, %r)" % (type(self).__name__,
+                               self.children[0],
+                               self.children[1])
+
+
+class TensorAdd(BinaryOp):
+    """Abstract SLATE class representing tensor addition."""
+
+    prec = 1
+    op = operator.add
+
+    @classmethod
+    def check_dimensions(cls, A, B):
+        """Checks the shapes of the tensors A and B before
+        attempting to perform tensor addition."""
+
+        if A.shape != B.shape:
+            dimension_error(A.shape, B.shape)
+
+    @classmethod
+    def get_bop_arguments(cls, A, B):
+        """Returns the appropriate arguments of the resulting
+        tensor via tensor addition."""
+
+        # Scalars distribute over sums
+        if isinstance(A, Scalar):
+            return B.arguments()
+        elif isinstance(B, Scalar):
+            return A.arguments()
+        return A.arguments()
+
+    @classmethod
+    def get_bop_integrals(cls, A, B):
+        """Returns the appropriate integrals of the resulting
+        tensor via tensor addition."""
+        return A.get_ufl_integrals() + B.get_ufl_integrals()
+
+
+class TensorSub(BinaryOp):
+    """Abstract SLATE class representing tensor subtraction."""
+
+    prec = 1
+    op = operator.sub
+
+    @classmethod
+    def check_dimensions(cls, A, B):
+        """Checks the shapes of the tensors A and B before
+        attempting to perform tensor subtraction."""
+
+        if A.shape != B.shape:
+            dimension_error(A.shape, B.shape)
+
+    @classmethod
+    def get_bop_arguments(cls, A, B):
+        """Returns the appropriate arguments of the resulting
+        tensor via tensor subtraction."""
+
+        # Scalars distribute over sums
+        if isinstance(A, Scalar):
+            return B.arguments()
+        elif isinstance(B, Scalar):
+            return A.arguments()
+        return A.arguments()
+
+    @classmethod
+    def get_bop_integrals(cls, A, B):
+        """Returns the appropriate integrals of the resulting
+        tensor via tensor subtraction."""
+        # TODO: Is it wasteful to declare an Negative tensor?
+        return A.get_ufl_integrals() + Negative(B).get_ufl_integrals()
```

Comment:
Is this plausibly even wrong?  Presumably `TensorSub` evaluates to `A - B`.  But if `B` is made from `Negative(B).integrals()` then my reading above is that this will actually evaluate `A + B`?